File: BOJ/2526.cpp
#include <cstdio>
#include <algorithm>
#include <vector>
using namespace std;
int n, p;
bool visited[1000005];
vector<int> a;
int main(){
	scanf("%d%d", &n, &p);
	visited[n] = true;
	int temp = n;
	int cnt = 0;
	a.push_back(n);
	while (1){
		temp = (temp*n) % p;
		if (visited[temp]){
			a.push_back(temp);
			break;
		}
		else {
			visited[temp] = true;
			a.push_back(temp);
		}
	}
	int la = a[a.size() - 1];
	for (int i = a.size() - 2; i >= 0; i--) {
		if (la == a[i]) break;
		else cnt++;
	}
	printf("%d\n", cnt+1);
	return 0;
}

File: BOJ(Java 공부용)/2526.java
import java.io.*;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken()), p = Integer.parseInt(st.nextToken());
        boolean[] visited = new boolean[1000001];
        ArrayList<Integer> v = new ArrayList<>();
        visited[n] = true;
        int temp = n, cnt = 0;
        v.add(n);
        while(true) {
            temp = (temp*n) % p;
            if(visited[temp]) {
                v.add(temp);
                break;
            }
            visited[temp] = true;
            v.add(temp);
        }
        int last = v.get(v.size()-1);
        for(int i=v.size()-2;i>=0;i--) {
            if(last == v.get(i)) break;
            cnt++;
        }
        System.out.println(cnt+1);
    }
}